src/README.md

```diff
@@ -0,0 +1,753 @@
+### libuv handles and requests
+
+Two central concepts when working with libuv are handles and requests.
+
+Handles are subclasses of the `uv_handle_t` “class”, and generally refer to
+long-lived objects that can emit events multiple times, such as network sockets
+or file system watchers.
+
+In Node.js, handles are often managed through a [`HandleWrap`][] subclass.
+
+Requests are one-time asynchronous function calls on the event loop, such as
+file system requests or network write operations, that either succeed or fail.
+
+In Node.js, requests are often managed through a [`ReqWrap`][] subclass.
+
+### Environment cleanup
+
+When a Node.js [`Environment`][] is destroyed, it generally needs to clean up
+any resources owned by it, e.g. memory or libuv requests/handles.
+
+<a id="cleanup-hooks"></a>
+#### Cleanup hooks
+
+Cleanup hooks are provided that run before the [`Environment`][]
+is destroyed. They can be added and removed through by using
+`env->AddCleanupHook(callback, hint);` and
+`env->RemoveCleanupHook(callback, hint);`, where callback takes a `void* hint`
+argument.
+
+Inside these cleanup hooks, new asynchronous operations *may* be started on the
+event loop, although ideally that is avoided as much as possible.
+
+Every [`BaseObject`][] has its own cleanup hook that deletes it. For
+[`ReqWrap`][] and [`HandleWrap`][] instances, cleanup of the associated libuv
+objects is performed automatically, i.e. handles are closed and requests
+are cancelled if possible.
+
+#### Closing libuv handles
+
+If a libuv handle is not managed through a [`HandleWrap`][] instance,
+it needs to be closed explicitly. Do not use `uv_close()` for that, but rather
+`env->CloseHandle()`, which works the same way but keeps track of the number
+of handles that are still closing.
+
+#### Closing libuv requests
+
+There is no way to abort libuv requests in general. If a libuv request is not
+managed through a [`ReqWrap`][] instance, the
+`env->IncreaseWaitingRequestCounter()` and
+`env->DecreaseWaitingRequestCounter()` functions need to be used to keep track
+of the number of active libuv requests.
+
+#### Calling into JavaScript
+
+Calling into JavaScript is not allowed during cleanup. Worker threads explicitly
+forbid this during their shutdown sequence, but the main thread does not for
+backwards compatibility reasons.
+
+When calling into JavaScript without using [`MakeCallback()`][], check the
+`env->can_call_into_js()` flag and do not proceed if it is set to `false`.
+
+## Classes associated with JavaScript objects
+
+### `MemoryRetainer`
+
+A large number of classes in the Node.js C++ codebase refer to other objects.
+Inheriting from the `MemoryRetainer` class enables objects in a heap snapshot
+to refer to instances of that class, and in turn enables that class to point
+to other objects as well, including native C++ types such as `std::string`
+and track their memory usage.
+
+This can be useful for debugging memory leaks.
+
```

Comment:
could we have a one liner about what an inheriting class should implement to track its memory?

Reply:
I’ve linked the header file, this one actually contains quite good documentation thanks to @joyeecheung